Source/Mutex+Protector.swift

```diff
@@ -0,0 +1,176 @@
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+
+/// A lock abstraction.
+private protocol Lock {
+    func lock()
+    func unlock()
+    func around<T>(_ closure: () -> T) -> T
+    func around(_ closure: () -> Void)
+}
+
+/// A `pthread_mutex` wrapper, inspired by ProcedureKit.
+final class Mutex: Lock {
+    private var mutex = pthread_mutex_t()
+
+    public init() {
+        let result = pthread_mutex_init(&mutex, nil)
+        precondition(result == 0, "Failed to create pthread mutex")
+    }
+
+    deinit {
+        let result = pthread_mutex_destroy(&mutex)
+        assert(result == 0, "Failed to destroy mutex")
+    }
+
+    fileprivate func lock() {
+        let result = pthread_mutex_lock(&mutex)
+        assert(result == 0, "Failed to lock mutex")
+    }
+
+    fileprivate func unlock() {
+        let result = pthread_mutex_unlock(&mutex)
+        assert(result == 0, "Failed to unlock mutex")
+    }
+
+    /// Execute a value producing closure while aquiring the mutex.
+    ///
+    /// - Parameter closure: The closure to run.
+    /// - Returns:           The value the closure generated.
+    func around<T>(_ closure: () -> T) -> T {
+        lock(); defer { unlock() }
+        return closure()
+    }
+
+    /// Execute a closure while aquiring the mutex.
+    ///
+    /// - Parameter closure: The closure to run.
+    func around(_ closure: () -> Void) {
+        lock(); defer { unlock() }
+        return closure()
+    }
+}
+
+/// An `os_unfair_lock` wrapper.
+@available (iOS 10.0, macOS 10.12, tvOS 10.0, watchOS 3.0, *)
+final class UnfairLock: Lock {
+    private var unfairLock = os_unfair_lock()
+
+    public init() { }
+
+    fileprivate func lock() {
+        os_unfair_lock_lock(&unfairLock)
+    }
+
+    fileprivate func unlock() {
+        os_unfair_lock_unlock(&unfairLock)
+    }
+
+    /// Execute a value producing closure while aquiring the lock.
+    ///
+    /// - Parameter closure: The closure to run.
+    /// - Returns:           The value the closure generated.
+    func around<T>(_ closure: () -> T) -> T {
+        lock(); defer { unlock() }
+        return closure()
+    }
+
+    /// Execute a closure while aquiring the lock.
+    ///
+    /// - Parameter closure: The closure to run.
+    func around(_ closure: () -> Void) {
+        lock(); defer { unlock() }
+        return closure()
+    }
+}
+
+/// A thread-safe wrapper around a value.
+final class Protector<T> {
+    private let lock: Lock = {
+        if #available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOS 3.0, *) {
+            return UnfairLock()
+        } else {
+            return Mutex()
+        }
+    }()
+    private var ward: T
+
+    init(_ ward: T) {
+        self.ward = ward
+    }
+
+    /// The contained value. Unsafe for anything more than direct read or write.
+    var unsafeValue: T {
+        get { return lock.around { ward } }
+        set { lock.around { ward = newValue } }
+    }
```

Comment:
Should we really name this `unsafe` when it's actually safe for direct reads and writes? Seems a bit scary when used throughout the codebase. Makes me think it's unsafe since it's part of the API.

Reply:
After a lot of thought, I've renamed the property to `directValue`. So while it isn't unsafe in the Swift sense of the word (i.e. could lead to undetected runtime memory corruption), I did want a name that made end users take notice. Fixed in [6907baf](https://github.com/Alamofire/Alamofire/pull/2290/commits/6907bafde7b8e9648147cd045f0bc102d58d8a0f).